src/inmanta/agent/handler.py

```diff
@@ -961,15 +1019,113 @@ def execute(self, ctx: HandlerContext, resource: TPurgeableResource, dry_run: bo
+          the intent to discover resources of the second type D already present on the network.
+        - D denotes the handler's Unmanaged Resource type. This is the type of the resources that have been
+          discovered and reported to the server. Objects of this type must be serializable.
+    """
+
+    def deploy(
+        self,
+        ctx: HandlerContext,
+        resource: R,
+        requires: abc.Mapping[ResourceIdStr, ResourceState],
+    ) -> None:
+        self.execute(ctx, resource)
+
+    def check_facts(self, ctx: HandlerContext, resource: resources.Resource) -> Dict[str, object]:
+        return {}
+
+    def pre(self, ctx: HandlerContext, resource: R) -> None:
+        """
+        Override this method to run custom logic before resource discovery.
+
+        :param ctx: Context object to report changes and logs to the agent and server.
+        :param resource: The Discovery Resource used to drive resource discovery.
+        """
+        pass
+
+    def post(self, ctx: HandlerContext, resource: R) -> None:
+        """
+        Override this method to run custom logic after resource discovery.
+
+        :param ctx: Context object to report changes and logs to the agent and server.
+        :param resource: The Discovery Resource used to drive resource discovery.
+        """
+        pass
+
+    @abstractmethod
+    def discover_resources(self, ctx: HandlerContext, discovery_resource: R) -> abc.Mapping[ResourceIdStr, D]:
+        """
+        Implement this method to define specific resource discovery logic. This method will be called
+        by the handler's `execute` method during deployment of the corresponding discovery resource
+        by the agent.
+        """
+        raise NotImplementedError
+
+    def execute(self, ctx: HandlerContext, resource: R, dry_run: bool = False) -> None:
+        """
+        Generic logic to perform during resource discovery. This method is called when the agent wants
+        to deploy the corresponding discovery resource. The default behaviour of this method is to call
+        the `discover_resources` method, serialize the returned values and report them to the server.
+        """
+
+        def _call_discovered_resource_create_batch(
+            discovered_resources: abc.Sequence[DiscoveredResource],
+        ) -> typing.Awaitable[Result]:
+            return self.get_client().discovered_resource_create_batch(
+                tid=self._agent.environment, discovered_resources=discovered_resources
+            )
+
+        try:
+            self.pre(ctx, resource)
+            # serialize resources and report to the server
+            discovered_resources: abc.Sequence[DiscoveredResource] = [
+                DiscoveredResource(discovered_resource_id=resource_id, values=json.loads(json_encode(values)))
+                for resource_id, values in self.discover_resources(ctx, resource).items()
```

Comment:
purely a matter of taste, but I would put the  `self.discover_resources(ctx, resource)` on a line of its own, as it is the main thing here. It is kind of hidden now.